core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -177,6 +190,41 @@ public RecoveryResponse recoverToTarget() throws IOException {
     }
 
     /**
+     * Determines if the source translog is ready for a sequence-number-based peer recovery. The main condition here is that the source
+     * translog contains all operations between the local checkpoint on the target and the current maximum sequence number on the source.
+     *
+     * @param translogView a view of the translog on the source
+     * @return {@code true} if the source is ready for a sequence-number-based recovery
+     * @throws IOException if an I/O exception occurred reading the translog snapshot
+     */
+    boolean isTranslogReadyForSequenceNumberBasedRecovery(final Translog.View translogView) throws IOException {
+        final long startingSeqNo = request.startingSeqNo();
+        final long endingSeqNo = shard.seqNoStats().getMaxSeqNo();
+        logger.trace("starting: [{}], ending: [{}}", startingSeqNo, endingSeqNo);
+        if (startingSeqNo - 1 <= endingSeqNo) {
```

Comment:
> I think we should log a warning if we have startSeqNo > endingSeqNo+1 and also assert it doesn't happen (so we fail in tests)

I do not think this is correct. Imagine a snapshot that restores a previous commit. After the snapshot is restored, a recovery will commence from the restored primary to the replica shards. The replica shards will still have files on disk that could have a commit point after the restored commit point. Their local checkpoint could exceed the maximum sequence number on the restored primary. The replicas will send a start recovery request where starting sequence number on the target is greater than maximum sequence number (plus one) on the source.